types/coin.go

```diff
@@ -182,43 +181,61 @@ func (coins Coins) String() string {
 	return out[:len(out)-1]
 }
 
-// IsValid asserts the Coins are sorted, have positive amount,
-// and Denom does not contain upper case characters.
-func (coins Coins) IsValid() bool {
+// Validate checks that the Coins are sorted, have positive amount, and Denom is valid. Otherwise,
+// it returns an error.
+func (coins Coins) Validate() error {
 	switch len(coins) {
 	case 0:
-		return true
+		return nil
+
 	case 1:
 		if err := ValidateDenom(coins[0].Denom); err != nil {
-			return false
+			return err
 		}
-		return coins[0].IsPositive()
+		if !coins[0].IsPositive() {
+			return fmt.Errorf("coin %s amount is not positive", coins[0])
+		}
+		return nil
+
 	default:
 		// check single coin case
-		if !(Coins{coins[0]}).IsValid() {
-			return false
+		if err := (Coins{coins[0]}).Validate(); err != nil {
+			return err
 		}
 
 		lowDenom := coins[0].Denom
+		seenDenoms := make(map[string]bool)
+		seenDenoms[lowDenom] = true
+
 		for _, coin := range coins[1:] {
-			if strings.ToLower(coin.Denom) != coin.Denom {
-				return false
+			if seenDenoms[coin.Denom] {
+				return fmt.Errorf("duplicate denomination %s", coin.Denom)
```

Comment:
this closes #6716 - could you update godoc for `ParseCoins` to mention an error is returned on invalid coins

Reply:
In hindsight, I think the `coin.Denom <= lowDenom` would catch dup denoms, but the original code is very poorly documented and seemingly redundant:

```
    // from NewCoins()

    // detect duplicate Denoms
    if dupIndex := findDup(newCoins); dupIndex != -1 {
        panic(fmt.Errorf("find duplicate denom: %s", newCoins[dupIndex]))
    }   

    if !newCoins.IsValid() {
        panic(fmt.Errorf("invalid coin set: %s", newCoins))
    }   
```
I'm pretty sure if coins are sorted it should always detect dups in `Validate`